README.md
# Hello MVC
This is an example of a very basic and stupid attempt to implement a design that somewhat resembles MVC, on a plain LAMP stack, with no external dependencies other than the PDO database driver for MariaDB, complete with fancy vanity routing and a basic RESTful CRUD API, made in a few hours instead of sleeping.

## Demo
[![image](https://user-images.githubusercontent.com/91397173/166156421-25b184c0-626b-47cd-b741-bd0154dc342d.png)](https://hello-mvc.chigeek.xyz/)

## Compose stack files
### `Dockerfile`
```Dockerfile
FROM php:7.4-apache AS base
RUN docker-php-ext-install mysqli pdo pdo_mysql && docker-php-ext-enable pdo_mysql
VOLUME /var/www/config

FROM base AS development
RUN mv "$PHP_INI_DIR/php.ini-development" "$PHP_INI_DIR/php.ini"

FROM base AS production
RUN mv "$PHP_INI_DIR/php.ini-production" "$PHP_INI_DIR/php.ini"
COPY controllers/ html/ lib/ routes/ views/ /var/www/
```
This dockerfile is a multi-stage build that has two three stages, which each build targets: `base`, `development`, and `production`. The `development` target builds and image that is lighter weight and does not include thr html files which will need to be bbind-mounted into the container as a volume. The `production` target builds a container suitable for production, with all of the html files added to the container as an immutable layer. The two targets also use different php.ini files since some options are helpful for debugging but should not be used in a production environment. The `base` stage contains layers that are commo to both targets. To build using docker cli, run `docker build --tag chigeekgreg/lampstack:development --target development.` for development, or `docker build --tag chigeekgreg/lampstack:production .` for production. When using Docker Buildkit, only the desired target stage is run and layers are cached to improve build times.

### `docker-compose.yml`
```yml
version: "2.4"
services:
  web:
    image: chigeekgreg/lampstack-web:production
    build:
      context: .
      target: production
    volumes:
      - ./config:/var/www/config
    ports:
      - "80:80"
    depends_on:
      db:
        condition: service_healthy
    restart: unless-stopped
  db:
    image: mariadb
    volumes:
      - db:/var/lib/mysql
      - ./sql:/docker-entrypoint-initdb.d
      - ./seed:/seed
    environment:
      MYSQL_RANDOM_ROOT_PASSWORD: "yes"
      MYSQL_USER: ${MYSQL_USER}
      MYSQL_PASSWORD: ${MYSQL_PASSWORD}
      MYSQL_DATABASE: ${MYSQL_DATABASE}
    healthcheck:
      test: ["CMD", "mysqladmin", "-u$MYSQL_USER", "-p$MYSQL_PASSWORD", "ping", "--silent"]
      interval: 5s
      start_period: 10s
      timeout: 5s
      retries: 2
    restart: unless-stopped
volumes:
  db:
```
This docker-compose file defines two services (`web` and `db`) and one named volume (`db`). Note that only port 80 is exposed. The database port does not need to be exposed because the two services can communicate over the named docker bridge network which compose creates by defatult.

### `docker-compose.override.yml` for development environment
```yml
# This file overrides docker-compose.yml for a development environment.
# Delete or rename this file to switch to production environment.
services:
  web:
    image: chigeekgreg/lampstack-web:development
    build:
      target: development
    volumes:
      - .:/var/www
```
The benefit of this override file is to support fast reloading of the content in the webroot by bind-mounting the html directory instead of packing it into the image. When `docker-compose up` is run, the directives from the override file are automatically merged over the configuration from the default `docker-compose.yml` file. To prevent that from happening, rename the override file to something else, like `docker-compose.dev.yml`. Then the production environment can be brought up with `docker compose up` and the development environment with `docker-compose -f docker-compose.yml -f docker-compose.dev.yml`.

### `.env`
```env
MYSQL_USER='lampstack'
MYSQL_PASSWORD='Changeme!'
MYSQL_DATABASE='lampstack'
```
The `.env` file is loaded by docker compose by default and can be used for variable replacement. It is possible to use other env file names and/or use muktiple env files that each map to specific services, but a single `.env` file allows environment variables to be shared by multiple services, and is the most common.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

Here we write upgrading notes.
 
### Added
 
### Changed
 
### Fixed


## [0.0.2] - 2022-04-25

GitHub metadata

### Added
- .github folder metadata
  - ISSUE_TEMPLATE
    - but_report.md - Don't send bug reports
    - feature_request.md - Don't request features
  - AUTHORS.md
  - CODEOWNERS - who to notify of a PR
  - CONTRIBUTING.md
  - CONTRIBUTORS.md
  - FUNDING.yml
  - pull_request_template.md
  - SECURITY.md
  - SUPPORT.md
- CHANGELOG.md - a standardized open source project CHANGELOG
- CITATION.ccf
- Jekyll
  - _.config.yml
- LICENSE

## [0.0.1] - 2022-04-22

Initial commit

### Added
- General project files
  - README.md - Project information
  - .gitignore - Ignore these files
- Basic lamp stack file structure
  - config/ - configuration data
    - config.php - database connection info
  - html/ - PHP files
    - index.php - home page
    - info.php - phpinfo()
    - styles.css - basic stylesheet
  - seed/ database init data
    - test.csv - test table
  - sql/ - database init scripts
    - 01-create-tables.sql
    - 02-load-sample-data.sql
- Docker support files
  - Dockerfile - build instructions
  - docker-compose.yml - sevices definition
  - docker-compose.override.yml - development environment overrides
  - .env - docker-compose env-file

[Unreleased]: https://github.com/olivierlacan/keep-a-changelog/compare/v0.0.2...HEAD
[0.0.2]: https://github.com/chigeekgreg/lampstack/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/chigeekgreg/lampstack/releases/tag/v0.0.1

controllers/test.php
<?php
include '../lib/controller.php';
class TestController extends Controller {

}
?>

routes/api.php
<?php
/**
 * Send an associative array as output. If the array contains a property named status, set the http status code to it.
 * 
 * @param array_result an associative array that may or may not contain a key named status
 * @param status HTTP status
 */
function send_json(array $array_result, ?int $status = NULL) {
    if(!is_null($status))
        http_response_code($status);
    elseif(isset($array_result['status']))
        http_response_code($array_result['status']);
    //else
    //    http_response_code(200);
    header('Content-Type: application/json; charset=utf-8');
    die(json_encode($array_result) . "\n");
}

include_once '../controllers/test.php';
$test = new TestController("test");
switch($_SERVER['REQUEST_METHOD']) {
    case 'GET':
        switch ($_SERVER['REQUEST_URI']) {
            case '/api/test':
            case '/api/test/':
                send_json($test->index());
                break;
        }
        break;
    case 'POST':
        switch ($_SERVER['REQUEST_URI']) {
            case '/api/test/new':
                send_json($test->insert($test->process_multipart_params($_POST)));
                break;
            case '/api/test/update':
                send_json($test->update($test->process_multipart_params($_POST)));
                break;
            case '/api/test/delete':
                send_json($test->delete($test->process_multipart_params($_POST)));
                break;
        }
        break;
}
?>

views/index.php
<?php
$base_url = (isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] === 'on' ? "https" : "http") . "://" . $_SERVER['HTTP_HOST'];
?>
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="color-scheme" content="dark light">
    <title>Hello MVC</title>
    <link rel="stylesheet" href="style.css">
</head>
<body>
    <h1>Hello MVC</h1>
    <p>
        This is an example of a very basic and stupid attempt to implement a design that somewhat resembles MVC,
        on a plain LAMP stack, with no external dependencies other than the PDO database driver for MariaDB,
        complete with fancy vanity routing and a basic RESTful CRUD API, made in a few hours instead of sleeping.
    </p>
    <hr>
    <h2>Page Routes</h2>
    <h3>GET</h3>
    <ul>
        <li>
            <details>
                <summary>/ - this view</summary>
                <strong>Example:</strong>
                <code>curl <?php echo $base_url; ?>/</code>
            </details>
        </li>
    </ul>
    <h3>POST</h3>
    <h2>API Routes</h2>
    <h3>GET</h3>
    <ul>
        <li>
            <details>
                <summary>/api/test - get the test table</summary>
                <strong>Example:</strong>
                <code>curl <?php echo $base_url; ?>/api/test</code>
            </details>
        </li>
    </ul>
    <h3>POST</h3>
    <ul>
        <li>
            <details>
                <summary>/api/test/new - create new items in the test table</summary>
                <strong>Example:</strong>
                <code>curl -F "title[]=foo title" -F "message[]=foo message" -F "title[]=bar title" -F "message[]=bar message" -X POST <?php echo $base_url; ?>/api/test/new</code>
            </details>
        </li>
        <li>
            <details>
                <summary>/api/test/update - update existing items in the test table, by ID</summary>
                <strong>Example:</strong>
                <code>curl -F "id[]=2" -F "title[]=frazz" -F "message[]=frazz message" -F "id[]=3" -F "title[]=matazz title" -F "message[]=matazz message" -X POST <?php echo $base_url; ?>/api/test/update</code>
            </details>
        </li>
        <li>
            <details>
                <summary>/api/test/delete - delete items from the test table, by any filter</summary>
                <strong>Example:</strong>
                <code>curl -F "id[]=2" -F "id[]=3" -X POST <?php echo $base_url; ?>/api/test/delete</code>
            </details>
        </li>
    </ul>
    <hr>
    <h2>Database test</h2>
    <!-- Run a simple `SELECT * FROM test` query and display the results. -->
<?php foreach ($array_result as $item):
    echo <<<HTML
    <div class="row">
        <h3>${item['title']}</h2>
        <p>${item['message']}</p>
    </div>
    HTML;
endforeach; ?>

</body>
</html>

routes/pages.php
<?php
/**
 * Send an associative array through a view as output. If the array contains a property named status, set the http status code to it.
 * 
 * @param view name of the view to use
 * @param array_result an associative array that may or may not contain a key named status
 * @param status HTTP status
 */
function send_view(string $view, array $array_result, ?int $status = NULL) {
    if(!is_null($status))
        http_response_code($status);
    elseif(isset($array_result['status']))
        http_response_code($array_result['status']);
    //else
    //    http_response_code(200);
    //header('Content-Type: text/html; charset=utf-8');
    if(!is_readable("../views/$view.php")) {
        http_response_code(500);
        die("Error: View $view does not exist.");
    }
    include_once "../views/$view.php";
    die("\n");
}
include_once '../controllers/test.php';
$test = new TestController("test");
switch($_SERVER['REQUEST_METHOD']) {
    case 'GET':
        switch ($_SERVER['REQUEST_URI']) {
            case '/':
                send_view('index', $test->index());
                break;
        }
        break;
    break;
}
?>

lib/controller.php
<?php
class Controller {

    /** @var db Database interface */
    protected $db;

    protected $name;

    /**
     * Controller class constructor
     */
    function __construct(?string $name = NULL) {
        include_once '../lib/db.php';
        $this->db = new DB();
        if(is_null($name))
            $this->name = preg_replace('/Controller$/', '', get_class($this));
        else
            $this->name = $name;
    }

    /** 
     * Process POST parameters into an associative array
     * 
     * @param postbody $_POST array
     */
    public function process_multipart_params(array &$postbody) {
        $objects = array();
        $params = array_keys($postbody);
        $count = -1;
        foreach($params as $param) {
            if($count == -1 || $count > count($postbody[$param]))
                $count = count($postbody[$param]);
        }
        for($i = 0; $i < $count; $i++) {
            $object = array();
            foreach($params as $param) {
                $object[$param] = $postbody[$param][$i];
            }
            $objects[] = $object;
        }
        return $objects;
    }

    /** 
     * Get all
     */
    public function index(): array {
        return $this->db->get_table($this->name);
    }

    /**
     * Insert multiple new objects
     * 
     * @param objects array of associative arrays.
     */
    public function insert(array $objects) {
        return $this->db->insert_many($this->name, $objects);
    }

    /**
     * Update multiple objects
     * 
     * @param objects array of associative arrays.
     */
    public function update(array $objects) {
        return $this->db->update_many($this->name, $objects);
    }

    /**
     * Update multiple objects
     * 
     * @param objects array of associative arrays.
     */
    public function delete(array $objects) {
        return $this->db->delete_many($this->name, $objects);
    }
}
?>

lib/db.php
<?php
class DB {

    /** @var pdo is in anstance of the PHP Data Objects class */
    protected $pdo;
    private $options = [
        PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
        PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC
    ];
    /**
     * DB class constructor
     */
    function __construct() {
        if(!is_readable('../config/config.php'))
            die("Can't open config file.");
        include '../config/config.php';
        if(!isset($config))
            die("Config not set");
        if(!isset($config['db']))
            die("Config:db not set");
        foreach(['host', 'name', 'user', 'pass', 'options'] as $setting) {
            if(!isset($config['db'][$setting]))
                die("Config:db:$setting not set\n");
        }
        $dsn = "mysql:host={$config['db']['host']};dbname={$config['db']['name']}";
        try {
            $this->pdo = new PDO($dsn, $config['db']['user'], $config['db']['pass']);
        } catch (PDOException $e) {
            die("Connection to database failed: ${$e->getMessage()}\n");
        }
    }

    /**
     * Query the database and return the result
     * 
     * This function takes a SQL query string and returns the result as a PDOStatement object
     */
    public function query(string $sql, ?array $args = NULL): PDOStatement {
        if(is_null($args) || empty($args)) {
            // If the $args is null or empty, return a simple PDOStatement::query()
            return $this->pdo->query($sql);
        } else {
            // Otherwise, use a prepared statement
            return $this->pdo->prepare($sql)->execute($args);
        }
    }

    /**
     * Get any table
     */
    public function get_table($tablename): array {
        return $this->query("SELECT * FROM $tablename")->fetchAll(PDO::FETCH_ASSOC);
    }

    /**
     * Safely insert many rows into a table
     * 
     * @param tablename table to insert into
     * @param rows row data
     * 
     * @return array result status
     * 
     */
    public function insert_many(string $tablename, array $rows): array {
        $results = [
            'status' => 200,
            'rows' => 0,
            'message' => 'OK'
        ];
        $columns = array_keys($rows[0]);
        $params = array();
        foreach($columns as $column) {
            $params[] = ":$column";
        }
        $sql = "INSERT INTO $tablename (" . implode(', ', $columns) . ") VALUES (" . implode(', ', $params) . ');';
        try {
            $statement = $this->pdo->prepare($sql);
        } catch (PDOException $e) {
            return [
                'status' => 500,
                'rows' => 0,
                'message' => 'Error. Could not prepare query. PDOException: ' . $e->getMessage()
            ];
        }
        foreach($rows as $row) {
            $data = array();
            foreach($columns as $column) {
                $data[":$column"] = $row[$column];
            }
            try {
                if($statement->execute($data))
                    $results['rows'] += $statement->rowCount();
                else
                    $results['message'] = 'Error. Could not execute insert for unknown reason.';
            } catch (PDOException $e) {
                $results['status'] = 500;
                $results['message'] = 'Error. Could not execute insert. PDOException: ' . $e->getMessage();
                return $results;
            }
        }
        return $results;
    }

    /**
     * Safely update many rows in a table by ID
     * 
     * @param tablename table to update
     * @param rows row data
     * 
     * @return array result status
     * 
     */
    public function update_many(string $tablename, array $rows): array {
        $results = [
            'status' => 200,
            'rows' => 0,
            'message' => 'OK'
        ];
        $columns = array_keys($rows[0]);
        $updates = array();
        foreach($columns as $column) {
            if($column != 'id')
                $updates[] = "$column = :$column";
        }
        $sql = "UPDATE $tablename SET " . implode(', ', $updates) . ' WHERE id = :id;';
        try {
            $statement = $this->pdo->prepare($sql);
        } catch (PDOException $e) {
            return [
                'status' => 500,
                'rows' => 0,
                'message' => 'Error. Could not prepare update. PDOException: ' . $e->getMessage()
            ];
        }
        foreach($rows as $row) {
            $data = array();
            foreach($columns as $column) {
                $data[":$column"] = $row[$column];
            }
            try {
                if($statement->execute($data))
                    $results['rows'] += $statement->rowCount();
                else
                    $results['message'] = 'Error. Could not execute update for unknown reason.';
            } catch (PDOException $e) {
                $results['status'] = 500;
                $results['message'] = 'Error. Could not execute update. PDOException: ' . $e->getMessage();
                return $results;
            }
        }
        return $results;
    }

    /**
     * Safely delete many rows in a table by ID
     * 
     * @param tablename table to delete from
     * @param rows row data
     * 
     * @return array result status
     * 
     */
    public function delete_many(string $tablename, array $rows): array {
        $results = [
            'status' => 200,
            'rows' => 0,
            'message' => 'OK'
        ];
        $columns = array_keys($rows[0]);
        $where = array();
        foreach($columns as $column) {
            $where[] = "$column = :$column";
        }
        $sql = "DELETE FROM $tablename WHERE " . implode(' AND ', $where) . ';';
        try {
            $statement = $this->pdo->prepare($sql);
        } catch (PDOException $e) {
            return [
                'status' => 500,
                'rows' => 0,
                'message' => 'Error. Could not prepare delete. PDOException: ' . $e->getMessage()
            ];
        }
        foreach($rows as $row) {
            $data = array();
            foreach($columns as $column) {
                $data[":$column"] = $row[$column];
            }
            try {
                if($statement->execute($data))
                    $results['rows'] += $statement->rowCount();
                else
                    $results['message'] = 'Error. Could not execute delete for unknown reason.';
            } catch (PDOException $e) {
                $results['status'] = 500;
                $results['message'] = 'Error. Could not execute delete. PDOException: ' . $e->getMessage();
                return $results;
            }
        }
        return $results;
    }
}
?>